# Chapter 1: Variables

*-- This course borrows heavily from 'Python for the Humanities', a course by Folgert Karsdorp and Maarten van Gompel with modifications and contributions by Mike Kestemont, Lars Wieneke and Bart Desmet*

-------------------------------

## First steps

Everyone can learn how to program and the best way to learn it is by doing it. This course on Processing Language with Python for people with no prior programming experience is extremely hands-on: you will have to write a lot of code yourself from the very beginning onwards. For writing Python code in this tutorial, you can use the many 'code blocks' you will encounter, such as the grey block immediately below. Place your cursor inside this block and press `ctrl+enter` to "run" or execute the code. Let's begin right away: run your first little program!

In [2]:
print("Hello world!")
print("Hello Jonas!")

Hello world!
Hello Jonas!


* Can you adapt the code in this box yourself and make it print your own name?

Apart from printing words to your screen, you can also use Python to do calculations. 

* Use the code block below to calculate how many minutes there are in seven weeks (hint: multiplication is done using the `*` symbol in Python.)


In [4]:
# insert your own code here!
print(7*7*24*60)


70560


Excellent! You have just written and executed your very first program! 

Note also that if a line starts with a hash this line is not part of the code but allows you to introduce additional comments: you will need these comments when you write more complex scripts!

Please make sure to run *every single one* of the following code blocks in the same manner - otherwise a lot of the examples won't properly work.

So far, we used only Python as a pretty minimalistic calculator, but there is more to discover. 

## Variables and values
Imagine that we want to store the number we just calculated so we can use it later. To do this we need to 'assign' a name to a value using the = symbol. 

In [6]:
x = 5
print("x =", x)

x = 5


If you vaguely remember your math classes in school, this should look familiar. It is basically the same notation with the name of the variable on the left, the value on the right, and the = sign in the middle. 

In the code block above, two things happen. First, we fill `x` with a value, in our case `5`. This variable x behaves pretty much like a box in which to store something. We write an `x` with a thick, black marker on it, so we can find it back later. We then print the contents of this box, using the `print()` command. 

* Now copy the outcome of your code calculating the number of minutes in seven weeks and assign this number to `x`. Run the code again.

The box metaphor for a variable goes a long way: in such a box you can put whatever value you want, e.g. the number of minutes in seven weeks. When you re-assign a variable, you remove the content of the box and  put something new in it. In Python, the term 'variable' refers to such a box, whereas the term 'value' refers to what is inside this box.

When we have stored values inside variables, we can do interesting things with these variables. You can, for instance, run the calculations in the block below to see the effect of the following five lines of code. Symbols like `=`, `+`, `-` , `/` `*` are called 'operators' in programming: they all provide a very basic functionality such as assigning values to variables or doing multiplication and subtraction.  

In [7]:
x = 2
print(x)
print(x * x)
print(x + x)
print(x / x)
print(x - 6)

2
4
4
1.0
-4


So far, we have only used a variable called `x`. Nevertheless, we are entirely free to change the names of our variables, as long as these names do not contain strange characters, such as spaces, dashes or punctuation marks. Underscores and numbers, however, are allowed inside names (so you cannot begin a variable name with them). In the following block, we assign the outcome of our calculation to a variable that has a more meaningful name than the abstract name `x`.

In [8]:
seconds_in_seven_weeks = 70560
print(seconds_in_seven_weeks)

70560


In Python we can also copy the contents of a variable *into another variable*, which is what happens in the second line of the code block below. Each variable will always contain the value that you *last* assigned to it:

In [9]:
first_number = 5
second_number = first_number
first_number = 3
print(first_number)
print(second_number)

3
5


Variables are also case sensitive, so accessing *first_number* is not the same as *First_number*

In [10]:
print(first_number)

3


In [11]:
print(First_number)

NameError: name 'First_number' is not defined

So far we have only assigned numbers such as `5` or `70560` to our variables. Such whole numbers are called 'integers' in programming, because they don't have any digits 'after the dot'.  Numbers that *do* have digits after the dot (e.g. `67.278` or `191.200`), are called 'floating-point numbers' in programming or simply 'floats'. Note that Python uses dots in floats, whereas some European languages use a comma here. Both integers and floats can be positive numbers (e.g. `70` or `4.36`) as well as negative numbers (e.g. `-70` or `4.36`). You can just as easily assign floats to variables:

In [12]:
some_float = 23.987
print(some_float)
some_float = -4.56
print(some_float)

23.987
-4.56


On the whole, the difference between integers and floats is of course important for divisions where you often end up with floats:

In [13]:
x = 5/2
print(x)

2.5


You will undoubtedly remember from your math classes in high school that there is something called 'operator precedence', meaning that multiplication, for instance, will always be executed before subtraction. In Python you can explicitly set the order in which arithmetic operations are executed, using round brackets. Compare the following lines of code:

In [14]:
nr1 = 10-2/4
nr2 = (10-2)/4
nr3 = 10-(2/4)
print(nr1)
print(nr2)
print(nr3)

9.5
2.0
9.5


Using the operators we have learned about above, we can change the variables in our code as many times as we want. We can assign new values to old variables, just like we can put new or more things in the boxes we already had. Say, for instance, that yesterday we counted how many books we have in our office and that we stored this count in our code as follows:

In [15]:
number_of_books = 100

Suppose that we buy a new book for the office today: we can now update our book count accordingly, by adding one to our previous count:

In [16]:
number_of_books = number_of_books + 1
print(number_of_books)

101


Here, the value of variable `number_of_books` is changed by manipulating its previously stored value. This works because assignment goes from *right to left*. In other words, the part on the right of the assignment sign `=` is executed first (so the old value of `number_of_books` is retrieved and `1` is added), and then the result of this operation is assigned to the variable name on the left.

Updates like these happen a lot. Python therefore provides a shortcut so you don't need to write the same variable name twice: you can write the same thing using `+=`.

In [17]:
number_of_books += 5
print(number_of_books)

106


This special shortcut (`+=`) is called an operator too. Apart from multiplication (`+=`), the operator has variants for subtraction (`-=`), multiplication (`*=`) and division (`/=`) too:

In [18]:
number_of_books -= 5
print(number_of_books)
number_of_books *= 2
print(number_of_books)
number_of_books /= 2
print(number_of_books)

101
202
101.0


----------

##### What we have learnt

To finish this section, here is an overview of the concepts you have learnt. Go through the list and make sure you understand all the concepts.

-  variable
-  value
-  assignment operator (`=`)
-  difference between variables and values
-  integers vs. floats
-  operators for multiplication (`*`), subtraction (`-`), addition (`+`), division (`/`)
-  the shortcut operators: `+=`, `-=`, `*=`, `/=`
-  `print()`


------

## Text strings

So far, we have only worked with variables that contain numbers (integers like `-5` or `72` or floats like `45.89` or `-5.609`). Note, however, that variables can also contain other things than numbers. It is likely that you will not be working primarily with numbers, but with texts. Quite naturally, this course will focus a lot on manipulating them. Have a look at the code block below, for instance. Here we put *text*, namely the title of a book, as a value inside the variable `book`. Then, we print what is inside the `book` variable.

In [20]:
book = "The Lord of the Flies"
print(book)
print(len(book))

The Lord of the Flies
21


Such a piece of text ("The Lord of the Flies") is called a 'string' in Python (cf. a *string* or *series* of characters). Strings in Python must always be enclosed with 'quotes' (either single or double quotes). Without those quotes, Python will think it's dealing with the name of some variable that has been defined earlier, because variable names never take quotes. The following distinction is confusing, but extremely important: variable names (*without* quotes) and string values (*with* quotes) look similar, but they serve a completely different purpose. Compare:

In [21]:
gender = "male"
male = 14
print(gender)
print ("male")
print(male)

male
male
14


Some of the arithmetic operators we saw earlier can also be used to do useful things with strings. Both the multiplication operator (`*`) and the addition operator (`+`) provide interesting functionality for dealing with strings, as the block below illustrates.


In [22]:
original_string = "bla"
new_string = 2*original_string
print(new_string)
new_string = new_string+"h"
print(new_string)

blabla
blablah


Adding strings together is called 'string concatenation' or simply 'concatenation' in programming. Use the block below to find out whether you can also use the shortcut `+=` operator for adding an 'h' to the variable `original_string`. Don't forget to check the result that gets printed!

In [26]:
original_string = "blabla"
# manipulate original_string to add an 'h'...
original_string += "h"
print(original_string)

blablah


Note that there is no subtraction operator for strings. Try the following code and you will see a typical error code (and learn that you will spend most of your programming time solving errors):

In [27]:
text = "blabla"
text -= "bla"

TypeError: unsupported operand type(s) for -=: 'str' and 'str'

Now we will write some code that defines variable file names. Let us say that the root of the filename is `Task1`. Now we will create a variable that defines the status of the file as `draft` or `revision` and we will add the `txt` extension. The result should be 'Task1_draft.txt' or 'Task1_revision.txt' 

In [36]:
file = "Task100"
status = "revision"
filename = file + "_" + status + ".txt"
print(filename)

Task100_revision.txt


Strings are called strings because they consist of a series (or 'string') of individual characters. We can access these individual characters in Python with the help of 'indexing', because each character in a string has a unique 'index'. To print the first letter of your name, you can type:

In [37]:
name = "Jonas"
first_letter = name[0]
print(first_letter)

J


![Indexes of the string Mr White starting with 0](images/string_index.png "Python strings and indexes")

Take a look at the string "Mr White". We use the index `0` to access the first character in the string. This might seem odd, but  remember that all indexes in Python start at zero. Whenever you count in Python, you start at `0` instead of `1`. Note that the space character gets an index too, namely 2, because for a computer, a space is a character like any other. This is something you will have to get used to!

Because you know the length of your name you can ask for its last letter: 

In [38]:
last_letter = name[4] # replace x with the last index of your name (tip: indexes start at 0)
print(last_letter)

s


It is rather inconvenient having to know how long our strings are if we want to find out what the last letter is. Python provides a simple way of accessing a string 'from the rear':

In [39]:
last_letter = name[-1]
print(last_letter)

s


To access the last character in a string you have to use the index [-1]. Alternatively, there is the `len()` command which returns the length of a string:

In [40]:
print(len(name))

5


Remember the `len()` command, because it is very useful to get information and solve problems. It does not only work on strings, but on other data types as well, as we will see in future chapters. 

Do you understand the following code block? Can you explain what is happening?

In [41]:
last_letter = name[len(name)-1]
print(last_letter)

s


Now can you write some code that defines a variable `penultimate_letter` and assigns to this variable the *second to last* letter of your name?

In [42]:
penultimate_letter = name[-2]
print(penultimate_letter)

a


You're starting to become a real expert at indexing strings. Now what if we would like to find out what the last two or three letters of our name are? In Python we can use so-called 'slice indices' or 'slices' for short. To find the first two letters of our name we type in:

In [43]:
first_two_letters = name[0:2]
print(first_two_letters)

Jo


The `0` index is optional, so we could just as well type in `name[:2]`. This says: take all characters of name until you reach index 2 (i.e. up to but excluding the third letter). We can also start at index 2 and leave the end index unspecified:

In [44]:
without_first_two_letters = name[2:]
print(without_first_two_letters)

nas


Because we did not specify the end index, Python continues until it reaches the end of our string. If we would like to find out what the last two letters of our name are, we can type in:

In [45]:
last_two_letters = name[-2:]
print(last_two_letters)

as


Slicing can be pretty confusing at first, mainly for two reasons:
* indices start at 0, so it's easy to make mistakes by starting your count from 1
* slices end *before* the last index you specify (i.e. from first index up to and *excluding* the last index)
It's normal to make mistakes. Just make sure that you check and debug your slices by printing them, and you will get the hang of it soon enough. You will need to use slices on other things than strings in later chapters, but the principles stay the same.

-------

## DIY

- Can you cut the filename into pieces and define two variables `name` and `extension` to which you assign the respective values?

In [52]:
filename = "Task01.txt"
name = filename[:-4]
extension = filename[-4:]
print(name)
print(extension)

Task01
.txt


- If you know that all the names end by two digits, can you then also separate the 'root' name and the number?

In [57]:
root = name[:-2]
number = name[-2:]
print(root)
print(number)

Task
01


- Given the following two words, can you write code that prints out the word *humanities* using only slicing and concatenation? (So, no quotes are allowed in your code.) Can you print out how many characters the word *humanities* contains?

In [60]:
word1 = "human"
word2 = "opportunities"
humanities = word1 + word2[-5:]
print(humanities)
print(len(humanities)) 

humanities
10


---------

## "Casting" variables

Above, we have already learned that each variable has a data type: variables can be strings, floats, integers, etc. Sometimes it is necessary to convert one type into the other. Consider this:

In [61]:
x = "5"
y = 2
print(x + y)

TypeError: must be str, not int

This should raise an error on your machine: does the error message gives you a hint as to why this does not work? *x* is a string, and *y* is an integer. Because of this, you cannot sum them. Luckily there exist ways to 'cast' variables from one type of variable into another type of variable.

- Do you understand the outcome of the following code?
- Can you comment in your own words on the effect of applying `int()` and `str()` to variables?

In [62]:
x = "5"
y = 2
print(x + str(y))
print(int(x) + y)

52
7


Casting is often necessary: it could be for example that you are reading numbers from a file (as strings) that you want to do calculations with. In that case you would need to cast the strings as integers (with `int()`) or as floats (with `float()`). Likewise, you sometimes need to write numbers to a file or print them to the screen together with text. In that case, you will often need to cast them as strings with `str()`, because Python will complain that it does not know how to combine a number with a string.

Let's go back to the filename example 'Task1' (with one digit to avoid further complications for the moment). Now we want to create a new filename that is one number higher than the original file name. How could we do that? Note: you will have to cast the value two times!

In [72]:
file_or = "Task1"
root = file_or[:-1]
number = file_or[-1:]
number_new = int(number) + 1
file_new = root + str(number_new)
print(file_new)

Task2


Other types of conversions are possible as well, and we will see a couple of them in the next chapters. Because variables can change data type anytime they want, we say that Python uses 'dynamic typing', as opposed to other more 'strict' languages that use 'strong typing'. You can check a variable's type using the `type()` command:

In [70]:
x = "5"
y = 2
print(type(x))
print(type(y))

<class 'str'>
<class 'int'>


## INPUT FIELDS

A pleasant way to introduce an interactive dimension in your scripts is by using input fields. Run the code below and see what happens.

In [73]:
naam = input ("What is your name?")
print('Hello, ' + naam + '!')

What is your name?Jonas
Hello, Jonas!


## COMMENTS

When you exchange code with fellow programmers (as you will often have to do in the real world), it is really helpful if you include some useful information about your scripts. It is also useful for you, to add information about what your code is meant to do. Have a look at the code block below and read about `commenting` on Python code in the comments:

In [74]:
# Have you noticed that everything behind the hashtag 
print("Something...") # on a line is ignored by your python interpreter?
print("and something else..") # this is really helpful to comment on your code!
"""Another way
of commenting on your code is via 
triple quotes -- these can be distributed over multiple """ # lines
print("Done.")

Something...
and something else..
Done.


So, how many ways are there to comment on your code in Python? 

----------

##### What we have learnt

To finish this section, here is an overview of what we have learnt. Go through the list and make sure you understand all the concepts.

-  string concatenation
-  index
-  slicing
- `zero-indexed` numbering
-  `len()`
-  type casting: `int()` and `str()`
-  `type()`
-  code commenting via hashtags and triple double quotes


-------

## Final Exercises Chapter 1

-  Ex. 1: Suppose the cover price of a book is 24.95 EUR, but bookstores get a 40 percent discount. Shipping costs 3 EUR for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies? As an extra challenge, try printing the result in a pretty fashion, using casting where necessary!

In [94]:
# your code goes here
number_of_copies = input("How many copies do you want? ")
price_books = (int(number_of_copies))*(24.95*0.6)
price_shipping = 3 + ((int(number_of_copies)-1)*0.75)
print("The cost of the books is €", price_books)
print("The shipping cost is €", price_shipping)
print("You have to pay €", price_books + price_shipping, "in total.")

How many copies do you want? 60
The cost of the books is € 898.1999999999999
The shipping cost is € 47.25
You have to pay € 945.4499999999999 in total.


-  Ex. 2: Can you identify and explain the errors in the following lines of code? Correct them please!

In [2]:
print("A message") # there is a point at the end
print("A message") #wrong use of quotation marks, the double ones are opened, but not closed
print('A message') #wrong use of quotation marks, there is a double one where it's not needed

A message
A message
A message


-  Ex. 3: When something is wrong with your code, Python will raise errors. Often these will be 'syntax errors' that signal that something is wrong with the form of your code (i.e. a `SyntaxError` like the one thrown in the previous exercise). There are also 'runtime errors' that signal that your code was in itself formally correct, but that something went wrong during the code's execution. A good example is the `ZeroDivisionError`. Try to make Python throw such a ZeroDivisionError! (Hint: it has to do with the fact that you cannot divide a number by zero)

In [3]:
# ZeroDivisionError
x = 6
y = 0
print(x/y)

ZeroDivisionError: division by zero

- Ex. 4: Write a program that assigns the result of `9.5 * 4.5 - 2.5 * 345.5 - 3.5` to a variable. Print this variable. Use round brackets to indicate 'operator precedence' and make sure that subtractions are performed before multiplications. Print the new result. When you convert the outcome to a string, how many characters does it count? Print that too.

In [5]:
# insert your code here
x = 9.5 * 4.5 - 2.5 * 345.5 - 3.5
print(x)
y = 9.5 * (4.5 - 2.5) * (345.5 - 3.5)
print(y)
print(len(str(y)))

-824.5
6498.0
6


-  Ex. 5: Define three variables `var1`, `var2` and `var3`. Calculate the average of these variables and assign it to `average`. Print the result in a fancy manner. Add three comments to this piece of code using three different ways.

In [11]:
# average
var1 = 10 # the average of the 3 variables is easy to calculate without Python
var2 = 15 
"""but since it's an exercise, we do it anyway"""
var3 = 5
average = (var1+var2+var3)/3
print("The average of var1, var2 and var3 is", average)

The average of var1, var2 and var3 is 10.0


-  Ex. 6: Write a little program that can compute the surface of a circle, using the variables `radius` and `pi=3.14159`. The formula is of course radius, multiplied by radius, multiplied by `pi`. Print the outcome of your program as follows: 'The surface area of a circle with radius ... is: ...'.

In [18]:
# circle code
radius = 6
pi = 3.14159
surface = radius*radius*pi
print("The surface area of a circle with radius ", radius, " is ", surface)

The surface area of a circle with radius  6  is  113.09724


To make things even more interactive, you can ask the user of the program to input the radius. Since this input is a string, you will need to cast it as an integer before doing calculations with it. Try adding your code from above to the code below and see if you can get it to work interactively.

In [20]:
# interactive circle code
radius = input("What is the radius? ")
radius = int(radius) # The interactive input is a string, so we need to cast it as an integer
pi = 3.14159
surface = radius*radius*pi
print("The surface area of a circle with radius ", radius, " is ", surface)

What is the radius? 8
The surface area of a circle with radius  8  is  201.06176


- Ex. 7: Get help online. There is one operator (like the ones for multiplication and subtraction) that we did not mention yet, namely the modulus operator `%`. Could you figure by yourself what it does when you place it between two numbers (e.g. 113 % 9)? If not, ask Google for help.

In [25]:
# try out the modulus operator!
x = 10 % 100 # 10 percent of 100 is 10
print(x)
y = 25 % 100 # 25 percent of 100 is 25
print(y)
z = 25 % 200 # 25 percent of 200 is 50
print(z)
x=14
y=3
print(x%y) # x/y=14/3=12, the rest is 2, the % sign calculates the rest when you devide the variable before the sign by the one after the sign

10
25
25
2


- Ex. 8: As happens very often, we make mistakes and Python raises errors. Correct all the errors in the following block of code. 

In [39]:
child1 = "Bert"
child2 = "An"
age_child1 = 5
age_child2 = 2

mother = "Marie"
father = "Jozef"

message1 = mother + " and " + father + " have two children, " + child1 + " and " + child2 + "." 
message2 = "Between the two children there is a difference of age of " + str(age_child1 - age_child2) + " years."
print(message1, message2)


Marie and Jozef have two children, Bert and An. Between the two children there is a difference of age of 3 years.


-----------------------

You've reached the end of Chapter 1! You can safely ignore the code block below -- it's only there to make the page prettier.

In [ ]:
from IPython.core.display import HTML
def css_styling():
    styles = open("styles/custom.css", "r").read()
    return HTML(styles)
css_styling()